proposals/22-04-localize-command.md

```diff
@@ -0,0 +1,296 @@
+# Localize Command
+
+**Authors**: 
+- annasong20
+
+**Reviewers**: 
+- natasha41575
+- KnVerey
+
+**Status**: implementable
+
+## Summary
+
+The `kustomize localize` command creates a “localized” copy of the target kustomization directory in which the layered
+kustomization files contain local, instead of remote, references to downloaded files. The command is part of an effort
+to enable `kustomize build` to run without network access.
+
+## Motivation
+
+monopole originally proposed the command `kustomize localize`
+in [this issue](https://github.com/kubernetes-sigs/kustomize/issues/3980).
+
+Users run `kustomize build` in many environments with limited network access. For example, CI/CD pipelines often only
+have access to the internal network. Server-side applications like Config Sync are concerned with the security
+vulnerabilities of git, which `kustomize build` uses to fetch remote files.
+
+These use cases would benefit from a kustomize solution that downloads all remote files that a `kustomize build` target
+references, into a copy of the target that references the downloaded files instead. `kustomize build` would then be able
+to run on the copy without a network dependency.
+
+This proposal nearly achieves the solution by downloading all remote files directly referenced by the target or by a
+recursively referenced kustomization file. The only remote files not covered by this proposal and still needed for
+`kustomize build` to run are those in KRM functions. However, KRM functions are third party to kustomize, and so will be
+out of scope for `kustomize localize`.
+
+**Goals:**
+1. This command should localize all remote files that a kustomization file directly references. This command will have
+   achieved this goal if, in the absence of remote input files to KRM functions, `kustomize build` can run on the
+   localized copy without network access.
+
+**Non-goals:**
+1. This command should not localize remote input files to KRM functions.
+1. This command should not serve as a package manager.
+
+## Proposal
+
+The command takes the following form:
+
+<pre>
+<b>kustomize localize</b> <ins>target</ins> <ins>newDir</ins>
+</pre>
+
+where the arguments are:
+
+* `target`: a directory with a top-level kustomization file that kustomize will localize; can be a path to a local
+  directory or a url to a remote directory
+* `newDir`: destination of the localized copy of `target`
+    * if `target` is local, `newDir` must be a directory name, as it will be located in the same directory as `target`
```

Comment:
Minor point to clarify, I think we have two options:

Option 1: I want to localize `dir/target`, so I run `kustomize localize dir/target localized-target`, which creates a localized version of `dir/target` under `dir`, in `dir/localized-target`. In this case, `newDir` needs to be just a name, and cannot be a filepath. 

Option 2:  If I want to localize a directory `dir/target`, I need to run `kustomize localize dir/target dir/localized-target`. If I run `kustomize localize dir/target localized-target`, I get an error saying that the filepath `localized-target` is not in the same directory as `dir/target`.

I slightly prefer the latter, WDYT?

Reply:
I'm also fine with either, but in Option 2 isn't it more work for the user to type the path? 

I guess Option 2 would help to check the user's understanding that `newDir` goes in the same directory as `target`? The user would also be giving explicit permission to write there?